### PR TITLE
[apr] bump to 1.7.5 and did some imporvements, availible from this version of the library.

### PR DESCRIPTION
**Pull request by rinrab**
Bump the library to version 1.7.5 and did some imporvements, availible from this version of the library.

* Making config to be official, since it was released in the library itself.
* Fix library linkage selection.
* Disable tests to improve the build performance.
* Install include files into the directory with 'apr-1' prefix.

### CODE REVIEW

File: ports/apr/portfile.cmake [hunk 1]
```diff
@@ -20,17 +19,24 @@ if (VCPKG_TARGET_IS_WINDOWS)
...
             -DINSTALL_PDB=OFF
             -DMIN_WINDOWS_VER=Windows7
             -DAPR_HAVE_IPV6=ON
+            -DAPR_INSTALL_INCLUDE_DIR=include/apr-1
             ${FEATURE_OPTIONS}
     )
 
     vcpkg_cmake_install()
-    vcpkg_cmake_config_fixup(PACKAGE_NAME unofficial-apr CONFIG_PATH share/unofficial-apr)
+    vcpkg_cmake_config_fixup(PACKAGE_NAME "apr" CONFIG_PATH "lib/cmake/apr")
```
Comment (dg0yt) [inline ports/apr/portfile.cmake:39]: Nit-pick:
```suggestion
    vcpkg_cmake_config_fixup(CONFIG_PATH "lib/cmake/apr")
```
Would it make sense to polyfill unofficial-apr-config.cmake, forwarding to apr with a warning, to mitigate migration woes?
(Already did that in some ports.

File: ports/apr/portfile.cmake [hunk 2]
```diff
@@ -20,17 +19,24 @@ if (VCPKG_TARGET_IS_WINDOWS)
...
     vcpkg_cmake_configure(
         SOURCE_PATH "${SOURCE_PATH}"
         OPTIONS
+            -DAPR_BUILD_STATIC=${APR_BUILD_STATIC}
+            -DAPR_BUILD_SHARED=${APR_BUILD_SHARED}
+            -DAPR_BUILD_TESTAPR=OFF
             -DINSTALL_PDB=OFF
             -DMIN_WINDOWS_VER=Windows7
             -DAPR_HAVE_IPV6=ON
+            -DAPR_INSTALL_INCLUDE_DIR=include/apr-1
```
Comment (dg0yt) [inline ports/apr/portfile.cmake:34]: Why is this necessary?
(No action required now.)

Comment (rinrab) [reply ports/apr/portfile.cmake]: Okay, reverted

File: ports/apr/portfile.cmake (hunk 1 above)
Comment (rinrab) [reply ports/apr/portfile.cmake]: I committed the suggestion (I didn't know already that it also works).

Also added unofficial-apr-config.cmake as suggested.

Comment (dg0yt) [reply ports/apr/portfile.cmake]: The default PACKAGE_NAME is the port name.

File: ports/apr/unofficial-apr-config.cmake [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+    add_library(unofficial::apr::aprapp-1 INTERFACE IMPORTED)
+    target_link_libraries(unofficial::apr::aprapp-1 INTERFACE apr::aprapp-1)
+else()
+    add_library(unofficial::apr::libapr-1 INTERFACE IMPORTED)
+    target_link_libraries(unofficial::apr::libapr-1 INTERFACE apr::libapr-1)
+
+    add_library(unofficial::apr::libaprapp-1 INTERFACE IMPORTED)
+    target_link_libraries(unofficial::apr::libaprapp-1 INTERFACE apr::libaprapp-1)
+endif()
```
Comment (dg0yt) [inline ports/apr/unofficial-apr-config.cmake:17]: Build-time triplet variable `VCPKG_LIBRARY_LINKAGE` is unknown at config use time, in project mode.
In addition, it is necessary to check if the target already exists.

```suggestion
if(TARGET apr::apr-1 AND NOT TARGET unofficial::apr::apr-1)
    add_library(unofficial::apr::apr-1 INTERFACE IMPORTED)
    target_link_libraries(unofficial::apr::apr-1 INTERFACE apr::apr-1)
elseif(TARGET apr::libapr-1 AND NOT TARGET unofficial::apr::libapr-1)
    add_library(unofficial::apr::libapr-1 INTERFACE IMPORTED)
    target_link_libraries(unofficial::apr::libapr-1 INTERFACE apr::libapr-1)
endif()
if(TARGET apr::aprapp-1 AND NOT TARGET unofficial::apr::aprapp-1)
    add_library(unofficial::apr::aprapp-1 INTERFACE IMPORTED)
    target_link_libraries(unofficial::apr::aprapp-1 INTERFACE apr::aprapp-1)
elseif(TARGET apr::libaprapp-1 AND NOT TARGET unofficial::apr::libaprapp-1)
    add_library(unofficial::apr::libaprapp-1 INTERFACE IMPORTED)
    target_link_libraries(unofficial::apr::libaprapp-1 INTERFACE apr::libaprapp-1)
endif()
```

I wish the upstream target name wouldn't change with library linkage, but I guess this ship has sailed.

Comment (rinrab) [reply ports/apr/unofficial-apr-config.cmake]: Ops, didn't notice. Pushed the fix as https://github.com/microsoft/vcpkg/pull/40742/commits/ce0ef6f49cf3a6741a97ec63257373f5db890ee0.

I think there should be the check for the target existing, because otherwise the target will be created, but will work incorrectly.

**Comment by JonLiu1993** [comment]
@rinrab, Thanks for your pr, when I tested the usage by `apr:x64-windows` , I get this error,please take a look:
````
1> CMake generation started for default configuration: 'x64-Debug'.
1> Found and using vcpkg toolchain file (F:/Feature-test/apr/scripts/buildsystems/vcpkg.cmake).
1> Command line: "C:\WINDOWS\system32\cmd.exe" /c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\test\source\repos\CMakeProject1\out\install\x64-Debug" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe"   -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" -DCMAKE_TOOLCHAIN_FILE="F:/Feature-test/apr/scripts/buildsystems/vcpkg.cmake" "C:\Users\test\source\repos\CMakeProject1" 2>&1"
1> Working directory: C:\Users\test\source\repos\CMakeProject1\out\build\x64-Debug
1> [CMake] -- The C compiler identification is MSVC 19.40.33814.0
1> [CMake] -- The CXX compiler identification is MSVC 19.40.33814.0
1> [CMake] -- Detecting C compiler ABI info
1> [CMake] -- Detecting C compiler ABI info - done
1> [CMake] -- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe - skipped
1> [CMake] -- Detecting C compile features
1> [CMake] -- Detecting C compile features - done
1> [CMake] -- Detecting CXX compiler ABI info
1> [CMake] -- Detecting CXX compiler ABI info - done
1> [CMake] -- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe - skipped
1> [CMake] -- Detecting CXX compile features
1> [CMake] -- Detecting CXX compile features - done
1> [CMake] CMake Warning at F:/Feature-test/apr/installed/x64-windows/share/unofficial-apr/unofficial-apr-config.cmake:1 (message):
1> [CMake]   find_package(unofficial-apr) is deprecated.
1> [CMake] 
1> [CMake]   Use find_package(apr) instead
1> [CMake] Call Stack (most recent call first):
1> [CMake]   F:/Feature-test/apr/scripts/buildsystems/vcpkg.cmake:859 (_find_package)
1> [CMake]   CMakeProject1/CMakeLists.txt:6 (find_package)
1> [CMake] 
1> [CMake] 
1> [CMake] -- Configuring done (2.0s)
1> [CMake] CMake Error at CMakeProject1/CMakeLists.txt:10 (target_link_libraries):
1> [CMake]   Target "test" links to:
1> [CMake] 
1> [CMake]     unofficial::apr::apr-1
1> [CMake] 
1> [CMake]   but the target was not found.  Possible reasons include:
1> [CMake] 
1> [CMake]     * There is a typo in the target name.
1> [CMake]     * A find_package call is missing for an IMPORTED target.
1> [CMake]     * An ALIAS target is missing.
1> [CMake] 
1> [CMake] 
1> [CMake] 
1> [CMake] -- Generating done (0.0s)
1> [CMake] CMake Generate step failed.  Build files cannot be regenerated correctly.
1> 'C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\test\source\repos\CMakeProject1\out\install\x64-Debug" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe"   -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" -DCMAKE_TOOLCHAIN_FILE="F:/Feature-test/apr/scripts/buildsystems/vcpkg.cmake" "C:\Users\test\source\repos\CMakeProject1" 2>&1"' execution failed with error: ''C:\WINDOWS\system32\cmd.exe' '/c "%SYSTEMROOT%\System32\chcp.com 65001 >NUL && "C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja"  -DCMAKE_BUILD_TYPE:STRING="Debug" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\test\source\repos\CMakeProject1\out\install\x64-Debug" -DCMAKE_C_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe" -DCMAKE_CXX_COMPILER:FILEPATH="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.40.33807/bin/Hostx64/x64/cl.exe"   -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES\MICROSOFT VISUAL STUDIO\2022\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" -DCMAKE_TOOLCHAIN_FILE="F:/Feature-test/apr/scripts/buildsystems/vcpkg.cmake" "C:\Users\test\source\repos\CMakeProject1" 2>&1"' returned with exit code: 1'.

````
CMakeLists.txt:
````
cmake_minimum_required(VERSION 3.5)

project(test)

find_package(unofficial-apr CONFIG REQUIRED)

add_executable(test CMakeProject1.cpp)

target_link_libraries(test PRIVATE unofficial::apr::apr-1 unofficial::apr::aprapp-1 unofficial::apr::libapr-1 unofficial::apr::libaprapp-1)
````

**Comment by rinrab** [comment]
This library has non-standard targets. By default, APR builds both static and shared configurations. The target names and library artifacts are `apr-1` for the static configuration and `libapr-1` for the shared configuration. However, either of these configurations can be disabled via options (which done by the port).

So, when you install the library for the `x64-windows` triplet, you only get the shared library (`libapr-1` and `aprapp-1`). This results in the non-existence of `apr::apr-1`, `apr::aprapp-1`, and their unofficial versions.

If you remove linking against `unofficial::apr::apr-1` and `unofficial::apr::aprapp-1` from your CMakeLists, it should work.

File: ports/apr/vcpkg.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "apr",
-  "version": "1.7.4",
+  "version": "1.7.5",
   "port-version": 1,
```
[the remark this answers is not in the document]
Comment (rinrab) [reply ports/apr/vcpkg.json]: Thanks! I forgot to remove it.

**Comment by JonLiu1993** [comment]
Tested usage successfully by `apr:x64-windows`:
````
apr provides CMake targets:

  # this is heuristically generated, and may not be correct
  find_package(apr CONFIG REQUIRED)
  target_link_libraries(main PRIVATE apr::libapr-1 apr::libaprapp-1)

  find_package(unofficial-apr CONFIG REQUIRED)
  target_link_libraries(main PRIVATE unofficial::apr::apr-1 unofficial::apr::aprapp-1 unofficial::apr::libapr-1 unofficial::apr::libaprapp-1)
````

**Review by JonLiu1993 — APPROVED** [review]

**Comment by dg0yt** [comment]
> ~~~
> target_link_libraries(main PRIVATE unofficial::apr::apr-1 unofficial::apr::aprapp-1 unofficial::apr::libapr-1 unofficial::apr::libaprapp-1)

Ouch. Why does this form of testing and feedback never stop?

Like all port with varying exported target names, this port should get a `usage` file with the `target_link_libraries(... $<IF:$<TARGET_EXISTS:...>,...,...>)` pattern.

**Comment by rinrab** [comment]
Opened another pull-request with adding the correct usage.
